### PR TITLE
debuginfo: Solve several problems when compiling Rust with `-g`

### LINKED ISSUE #19848: Building rustc with -g hits LLVM assert
```
$ RUSTFLAGS="-g" make -j8 rustc-stage1
...
rustc: x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore
rustc: /home/rustbuild/src/rust-buildbot/slave/snap3-linux/build/src/llvm/lib/IR/Instructions.cpp:1086: void llvm::StoreInst::AssertOK(): Assertion `getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"' failed.
make: *** [x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.core] Aborted
```

Backtrace from core file:

```
#0  0x00007f506ba0a877 in raise () from /lib64/libc.so.6
#1  0x00007f506ba0bf68 in abort () from /lib64/libc.so.6
#2  0x00007f506ba037d6 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007f506ba03882 in __assert_fail () from /lib64/libc.so.6
#4  0x00007f506e118c61 in llvm::StoreInst::AssertOK() ()
#5  0x00007f506d7c6666 in llvm::IRBuilder<true, llvm::ConstantFolder, llvm::IRBuilderDefaultInserter<true> >::CreateStore(llvm::Value*, llvm::Value*, bool) ()
#6  0x00007f506d261e62 in trans::builder::Builder$LT$$x27a$C$$x20$x27tcx$GT$::store::h30422c6d66c8362fMar ()
#7  0x00007f506d277697 in trans::base::trans_closure::hd73aafc90fa3ac9cVfu ()
#8  0x00007f506d211d82 in trans::expr::trans_rvalue_dps_unadjusted::h9838b52bc02d24d9cgj ()
#9  0x00007f506d20d688 in trans::expr::trans_unadjusted::h7b542949e8fd1f0aUCi ()
#10 0x00007f506d1c736c in trans::expr::trans::h6758cf65ffe4e433IQh ()
#11 0x00007f506d2068c0 in trans::callee::trans_args::h87945a0a00d776c99oh ()
#12 0x00007f506d1cea5d in trans::callee::trans_call_inner::h3407dfe88b63d12913g ()
#13 0x00007f506d20f8ef in trans::expr::trans_rvalue_dps_unadjusted::h9838b52bc02d24d9cgj ()
#14 0x00007f506d1c60f8 in trans::expr::trans_into::h3e5c3cd16b2b4e12WMh ()
#15 0x00007f506d1c6df4 in trans::controlflow::trans_block::hf81d152713271e77FYd ()
#16 0x00007f506d27842f in trans::base::trans_closure::hd73aafc90fa3ac9cVfu ()
#17 0x00007f506d1b94ce in trans::base::trans_fn::h8314c0146f176259cru ()
#18 0x00007f506d1b5007 in trans::base::trans_item::hdac7426481f8746aeMu ()
#19 0x00007f506d279c49 in trans::base::trans_mod::h03145c37fc751d9cJRu ()
#20 0x00007f506d1b4e7f in trans::base::trans_item::hdac7426481f8746aeMu ()
#21 0x00007f506d279c49 in trans::base::trans_mod::h03145c37fc751d9cJRu ()
#22 0x00007f506d27ed07 in trans::base::trans_crate::hff0a239bca25518eoIv ()
#23 0x00007f506cca175d in driver::phase_4_translate_to_llvm::h1d8e5fa121da5744BCa ()
#24 0x00007f506cc88b98 in driver::compile_input::h99875561ae679834pba ()
#25 0x00007f506cd23d38 in run_compiler::h68c8d9fcc09b1d69kYb ()
#26 0x00007f506cd2262c in run::closure.21400 ()
#27 0x00007f506cd33f3b in task::TaskBuilder::try_future::closure.22849 ()
#28 0x00007f506e8e18f3 in task::TaskBuilder::spawn_internal::closure.30601 ()
#29 0x00007f506e8fc732 in task::Task::spawn::closure.5727 ()
#30 0x00007f506e8fee9c in rust_try_inner ()
#31 0x00007f506e8fee86 in rust_try ()
#32 0x00007f506e8fc7d3 in unwind::try::h67138f1a9143bb4fTyc ()
#33 0x00007f506e8fc5dc in task::Task::run::hd7a7ebf54ed491e4fKb ()
#34 0x00007f506e8fc228 in task::Task::spawn::closure.5703 ()
#35 0x00007f506e8fd655 in thread::thread_start::h69086febf2f5cecdw1b ()
#36 0x00007f506c6c3ee5 in start_thread () from /lib64/libpthread.so.0
#37 0x00007f506bac9b8d in clone () from /lib64/libc.so.6
```

This looks similar to #17257, but I wasn't able to reproduce crashes on some of the examples in there, so I'm opening a separate issue for now.

### LINKED ISSUE #20798: Error building rust with -g
```
rustc: /home/rustbuild/src/rust-buildbot/slave/snap3-linux/build/src/llvm/lib/IR/DebugInfo.cpp:159: uint64_t llvm::DIVariable::getPieceOffset() const: Assertion `isVariablePiece()' failed.
Aborted (core dumped)
make: *** [x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rustc_typeck] Error 134
```

(I was running ` RUSTFLAGS="-g" CFG_RUSTC_FLAGS="-g" make -j4`, since I forgot which flag works)

I get the same error even after a make clean.

### PR DESCRIPTION
**Pull request by michaelwoerister**
Resolves some issues caused by the recent LLVM update (which itself solved some issues).

Closes #19848
Closes #20798

### CODE REVIEW

**Comment by brson** [comment]
r=me

**Comment by michaelwoerister** [comment]
rebased

File: src/librustc_trans/trans/_match.rs [hunk 1]
```diff
@@ -273,29 +277,39 @@ impl<'a, 'tcx> Opt<'a, 'tcx> {
...
-            SliceLengthGreaterOrEqual(prefix, suffix) => {
+            SliceLengthGreaterOrEqual(prefix, suffix, _) => {
                 LowerBound(Result::new(bcx, C_uint(ccx, prefix + suffix)))
             }
         }
     }
+
+    fn debug_loc(&self) -> DebugLoc {
+        match *self {
+            ConstantValue(_,debug_loc)                 |
```
Comment (taralx) [inline src/librustc_trans/trans/_match.rs]: Oops, missed a space after comma here.

This is why we need a rustfmt. :)

**Pull request merged**